runtime/parachains/src/paras_inherent.rs

```diff
@@ -927,6 +924,9 @@ fn apply_weight_limit<T: Config + inclusion::Config>(
 		rng,
 		bitfields.clone(),
 		vec![],
+		// bitfields can lead to enacting a candidate; however we don't have a good way of
+		// accounting for that when tracking individual bitfield weight, thus the weight here likely
+		// ends up being an underestimate.
```

Comment:
Underestimates are generally preferable. I am leaning towards doing the accumulation of bitfields and actually detecting which one is going to be enacted, given reasonable additional cost.

Reply:
The main issue I see with this approach is that we do PendingAvailaibility + new bitfields (in other words pre-existing validity votes + new validity votes) to calculate which candidates to enact.

Specifically, prior to accounting for the new bitfields we get the pending availability of every para id (in other words availability votes from prior blocks):

https://github.com/paritytech/polkadot/blob/64e1616fcc03896bc4eb0ff213dd226e8a1c3f4b/runtime/parachains/src/inclusion.rs#L314-L319

then for each new bitfield, we increment the availability votes for the corresponding para candidate:

https://github.com/paritytech/polkadot/blob/64e1616fcc03896bc4eb0ff213dd226e8a1c3f4b/runtime/parachains/src/inclusion.rs#L345-L350

and finally if the accumulation of availability votes put a candidate over the threshold we enact it:

https://github.com/paritytech/polkadot/blob/64e1616fcc03896bc4eb0ff213dd226e8a1c3f4b/runtime/parachains/src/inclusion.rs#L367

Reply:
What we could do is: if the new votes add up to x% of the threshold for it to be enacted, then we include the enactment weight because we think its highly likely to be enacted. If we expect most candidates to get all their votes in one block this should be ok for the average case. However I still see an issue for the edge case where in block N we get `threshold - x% + 1` votes and then in block N+1 we get `x% - r` votes - so the candidate gets enacted in block N+1 but we didn't think it would

Reply:
>  I am leaning towards doing the accumulation of bitfields and actually detecting which one is going to be enacted, given reasonable additional cost.

Given my previous comment, do you think its reasonable to try the above suggestion in this PR?

Reply:
https://github.com/paritytech/polkadot/issues/4513